### PR TITLE
all_metrics and all_companies

### PR DESCRIPTION
**Pull request by xithan**

### CODE REVIEW

File: mod/01_rating/set/abstract/all_metric_values.rb [hunk 1]
```diff
@@ -61,9 +61,7 @@ def count _params={}
 format do
   def search_results _args={}
     @search_results ||= begin
-      sort_by, sort_order = card.sort_params
-      all_results = sorted_result sort_by, sort_order, num?
```
Comment (xithan) [inline mod/01_rating/set/abstract/all_metric_values.rb:65]: no reason to pass all the stuff around if we fetch it from Env.params

File: mod/01_rating/set/type_plus_right/metric/all_metric_values.rb [hunk 2]
```diff
@@ -1,4 +1,4 @@
-include_set Abstract::AllValues
+include_set Abstract::AllMetricValues
```
Comment (xithan) [inline mod/01_rating/set/type_plus_right/metric/all_metric_values.rb:1]: I renamed it but now we use that abstract set also for lists of companies and metrics. We probably should come up with a better name

Comment (xithan) [reply mod/01_rating/set/type_plus_right/metric/all_metric_values.rb:1]: Maybe SortAndFilter is the right name

File: mod/01_rating/set/abstract/02_sort_and_filter/filter.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+def filter_by_value values
+  unit =
+    if Env.params["value"].present?
+      Env.params["value"]
+    else
+      "exists"
+    end
+  return values.empty? if unit == "none"
+  return !values.empty? if unit == "exists"
+  within_recent? unit, values
```
Comment (ethn) [inline mod/01_rating/set/abstract/02_sort_and_filter/filter.rb]: case statement?

File: mod/01_rating/set/abstract/02_sort_and_filter/filter.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+end
+
+def filter_by_year values
+  return true unless Env.params["year"].present?
+  values.any? { |v| v["year"] == Env.params["year"] }
```
Comment (ethn) [inline mod/01_rating/set/abstract/02_sort_and_filter/filter.rb]: for all these, if filter_by_X returns true, then we don't filter it, right?  That keeps throwing me.

File: mod/01_rating/set/abstract/02_sort_and_filter/sort.rb [hunk 5]
```diff
@@ -47,6 +47,16 @@ def unknown_value? value
...
+
+  # FIXME: hack to use "name" as query param
+  def path args={}
+    # the filter name cause conflict with the path name
+    return super unless args.delete(:replace_name)
+    if (name = args.delete(:name))
+      args[:_name_] = name
+    end
+    super(args).gsub("_name_=", "name=")
+  end
```
Comment (ethn) [inline mod/01_rating/set/abstract/02_sort_and_filter/sort.rb:59]: when my latest wagn pull request gets merged, this should be unnecessary

Comment (ethn) [reply mod/01_rating/set/abstract/02_sort_and_filter/sort.rb:59]: I'm guessing we want to get rid of `:replace name` somewhere...

File: mod/01_rating/set/abstract/02_sort_and_filter/sort.rb [hunk 6]
```diff
@@ -56,34 +66,26 @@ def latest_year_value values
   # @option args [String] :order
   # @option args [String] :class additional css class
   def sort_link text, args
+    path = { offset: offset, sort_order: args[:order],
+             limit: limit,   sort_by:    args[:sort_by],
+             replace_name: true }
```
Comment (ethn) [inline mod/01_rating/set/abstract/02_sort_and_filter/sort.rb:71]: here's the replace_name we won't want.

File: mod/caching/set/type_plus_right/wikirate_topic/all_company.rb [hunk 7]
```diff
@@ -1,8 +1,8 @@
-include Card::CachedCount
+include_set Abstract::SolidCache, cached_format: :json
 
 # refresh the topic+all company if source's company changed
 ensure_set { TypePlusRight::Source::WikiRateCompany }
-recount_trigger TypePlusRight::Source::WikirateCompany do |changed_card|
+cache_update_trigger TypePlusRight::Source::WikirateCompany do |changed_card|
```
Comment (ethn) [inline mod/caching/set/type_plus_right/wikirate_topic/all_company.rb:5]: so nice.  these changes are going to start me on a path back to sanity.

File: mod/01_rating/set/abstract/all_metric_values.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        #{wql_to_identify_related_metric_values}
+      },
+      "right":"value",
+      "limit":0
+    })
+end
+
+format :json do
+  view :core do |_args|
+    mvc = MetricValuesHash.new card.left
```
Comment (ethn) [inline mod/01_rating/set/abstract/all_metric_values.rb]: my brain said:

> does this mean mvh?
> maybe the _c_ is something different.
> maybe it's _card_.
> wait, that doesn't make sen...
> squirrel!

File: mod/01_rating/set/type_plus_right/metric/all_metric_values.rb (hunk 2 above)
Comment (ethn) [reply mod/01_rating/set/type_plus_right/metric/all_metric_values.rb:1]: ooh, we might be a small abstraction away from a very broadly usable mod!

**Review by ethn — APPROVED** [review]
makes me so happy to see these things getting more orderly.  thank you!

File: mod/01_rating/set/abstract/all_metric_values.rb (hunk 8 above)
Comment (xithan) [reply mod/01_rating/set/abstract/all_metric_values.rb]: I think it was MetricValuesCache first and I renamed it later

File: mod/01_rating/set/abstract/02_sort_and_filter/filter.rb (hunk 4 above)
Comment (xithan) [reply mod/01_rating/set/abstract/02_sort_and_filter/filter.rb]: yeah, noticed that too but with all the other changes I decided not to touch that. I renamed the method that calls all these methods "pass_filter?" to make it a bit clearer

File: mod/01_rating/set/abstract/02_sort_and_filter/filter.rb (hunk 3 above)
Comment (xithan) [reply mod/01_rating/set/abstract/02_sort_and_filter/filter.rb]: ✔️

File: mod/01_rating/set/abstract/02_sort_and_filter/filter.rb [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  %w(name industry project)
+end
+
+def keep_if field, args={}
+  filter =
+    Env.params[field.to_s].present? ? Env.params[field.to_s] : args[:default]
+  return true unless filter
+  return true if args[:options] && field.is_a?(Array) &&
+                 field.size == args[:options]  # all options selected
+  yield filter
```
Comment (ethn) [inline mod/01_rating/set/abstract/02_sort_and_filter/filter.rb:106]: elegant stuff.

**Pull request merged**